<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zheng Li" />

<meta name="date" content="2022-02-06" />

<title>Simulation</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BASS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Introduction</a>
</li>
<li>
  <a href="about.html">Software tutorial</a>
</li>
<li>
  <a href="simu.html">Simulation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Real data analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="STARmap.html">STARmap (Wang et al., 2018)</a>
    </li>
    <li>
      <a href="MERFISH.html">MERFISH (Moffitt et al., 2018)</a>
    </li>
    <li>
      <a href="DLPFC.html">DLPFC (Maynard et al., 2021)</a>
    </li>
  </ul>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zhengli09//BASS-analysis">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulation</h1>
<h4 class="author">Zheng Li</h4>
<h4 class="date">2022-02-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-02-27
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>BASS-analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed0code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(0)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed0code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(0)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzhengli09BASSanalysistree4ec8df1262087c7c78f682e56014a7087a756d4ftargetblank4ec8df1a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/zhengli09//BASS-analysis/tree/4ec8df1262087c7c78f682e56014a7087a756d4f" target="_blank">4ec8df1</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomzhengli09BASSanalysistree4ec8df1262087c7c78f682e56014a7087a756d4ftargetblank4ec8df1a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/zhengli09//BASS-analysis/tree/4ec8df1262087c7c78f682e56014a7087a756d4f" target="_blank">4ec8df1</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/simu.Rmd</code>) and HTML (<code>docs/simu.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zhengli09//BASS-analysis/blob/4ec8df1262087c7c78f682e56014a7087a756d4f/analysis/simu.Rmd" target="_blank">4ec8df1</a>
</td>
<td>
zhengli09
</td>
<td>
2022-02-27
</td>
<td>
Publish the introduction, simulation analysis and STARmap analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This page provides code to reproduce all the simulation results for BASS. Hopefully, this can also serve as a resource for building up your simulation studies and evaluate your statistical methods.</p>
</div>
<div id="simulation-study" class="section level1">
<h1>Simulation study</h1>
<div id="generate-simulated-data" class="section level2">
<h2>Generate simulated data</h2>
<p>The simulation study relies on a few R packages and python softwares (e.g. SpaGCN and FICT), which need to be installed beforehand. The specific packages and all necessary functions to conduct the simulation study can be found at <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/code/">simulation</a>.</p>
<pre class="r"><code>source(&quot;code/simu_utils.R&quot;)
# data for inferring parameters for splatter
load(&quot;data/starmap_mpfc.RData&quot;) # starmap_cnts, starmap_info
cnts &lt;- starmap_cnts[[&quot;20180417_BZ5_control&quot;]]
info &lt;- starmap_info[[&quot;20180417_BZ5_control&quot;]]
xy &lt;- as.matrix(info[, c(&quot;x&quot;, &quot;y&quot;)])
starmap &lt;- list(cnts = cnts, info = info)

# we perform 50 simulation replicates for each setting
# The random seeds used for each replicate are generated as below
NREP &lt;- 50
set.seed(0)
seeds &lt;- sample(20201230, NREP)</code></pre>
<p>We generate the simulated data under the baseline simulation setting. Refer to the <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/data/params.txt">params</a> file for a complete list of simulation settings.</p>
<pre class="r"><code># baseline settings for all four scenarios for the 
# single tissue section analysis (L = 1)
scenarios &lt;- 1:4 # simulation scenarios
C &lt;- 4 # number of cell types
R &lt;- 4 # number of spatial domains
J &lt;- 200 # number of genes
de_prop &lt;- 0.2 # proportion of DE genes for each cell type
de_facLoc &lt;- 1.1 # DE gene strength
L &lt;- 1 # number of tissue sections

# illustrate the data at the first replicate
rep &lt;- 1
sim_dats &lt;- lapply(scenarios, function(scenario){
  simu(starmap = starmap, scenario = scenario, C = C, J = J, L = L, 
       batch_facLoc = 0, de_prop = de_prop, de_facLoc = de_facLoc, 
       de_facScale = 0.4, sim_seed = seeds[rep], debug = FALSE)
})</code></pre>
<pre class="r"><code># The simulated data include a list of L gene expression count matrices,
# a list of L true cell type label vectors for evaluation purpose, and 
# the random seed used for generating the data
scenario &lt;- 3
sim_dat &lt;- sim_dats[[scenario]]
# gene expression count matrix of the first tissue section under scenario 3
sim_dat[[1]][[1]][1:5, 1:5]</code></pre>
<pre><code>      Cell1 Cell2 Cell3 Cell4 Cell5
Gene1     0     0     0     0     0
Gene2    16     0     0     0     0
Gene3   128     3     2     5     0
Gene4    17     0     0     1     2
Gene5     0     0     0     2     0</code></pre>
</div>
<div id="visualize-of-the-simulated-data" class="section level2">
<h2>Visualize of the simulated data</h2>
<p>We visualize the cell type distributions under scenarios I-IV. You can refer to <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/code/viz.R">visualization</a> for some useful plotting functions or you can write your own code for plotting.</p>
<pre class="r"><code>library(cowplot)
source(&quot;code/viz.R&quot;)
p &lt;- lapply(scenarios, function(scenario){
  plotClusters(xy, labels = sim_dats[[scenario]][[2]][[1]],
               title = paste(&quot;Scenario&quot;, c(&quot;I&quot;, &quot;II&quot;, &quot;III&quot;, &quot;IV&quot;)[scenario]))
})
legend &lt;- get_legend(p[[1]] + 
  theme(legend.position = &quot;bottom&quot;) +
  guides(color = guide_legend(title = &quot;Cell type&quot;)))
p &lt;- plot_grid(plotlist = p, ncol = 2)
plot_grid(p, legend, ncol = 1, rel_heights = c(1, 0.1))</code></pre>
<p><img src="figure/simu.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="running-all-methods" class="section level2">
<h2>Running all methods</h2>
<p>We have wrapped up code for running different methods into different functions. Those functions take in simulated data and output adjusted random index (ARI) measuring the estimate cell type or spatial domain labels with the underlying truth. Functions for running different methods can be found at <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/code/">simulation</a>. We take the simulated data from scenario 3 as an example to illustrate running different methods.</p>
<pre class="r"><code>scenario &lt;- 3</code></pre>
<div id="run-bass" class="section level3">
<h3>Run BASS</h3>
<pre class="r"><code>BASS_out &lt;- run_BASS(sim_dats[[scenario]], xy, &quot;ACCUR_EST&quot;, beta = 1, C, R, 
                     init_method = &quot;kmeans&quot;, cov_struc = &quot;EEE&quot;)</code></pre>
<pre><code>***************************************
  Bayesian Analytics for Spatial Segmentation (BASS)
  Authors: Zheng Li, Xiang Zhou
  Affiliate: Department of Biostatistics, University of Michigan
  INPUT INFO:
    - Number of samples: 1 
    - Number of spots/cells: 1049 
    - Number of genes: 200 
    - Potts interaction parameter estimation approach: ACCUR_EST 
    - Variance-covariance structure of gene expression features: EEE 
  To list all hyper-parameters, Type listAllHyper(BASS_object)
***************************************
***** Log-normalize gene expression data *****
***** Exclude genes with 0 expression *****
***** Reduce data dimension with PCA *****

    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                17.565               Converged (3 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 19.3 seconds. 

    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                10.228               Converged (2 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 11.8 seconds. </code></pre>
<pre class="r"><code>BASS_out</code></pre>
<pre><code>$c_ari
[1] 0.9957436

$z_ari
[1] 0.9190263

$pi_est
          [,1]      [,2]      [,3]      [,4]
[1,] 0.5411765 0.0000000 0.2703297 0.2582583
[2,] 0.2000000 0.4829545 0.0000000 0.2762763
[3,] 0.2588235 0.2272727 0.4703297 0.0000000
[4,] 0.0000000 0.2897727 0.2593407 0.4654655

$mse_pi
[1] 0.05127128</code></pre>
</div>
<div id="run-hmrf" class="section level3">
<h3>Run HMRF</h3>
<p>Please note that running HMRF requires you to specify a path for python and will generate intermediate and final results into a specified folder. Please modify those directories accordingly in the <code>run_HMRF</code> function if you want to use the function for your own analysis.</p>
<pre class="r"><code># Run HMRF
# Note that case and rep parameters below are used to generate a unique
# temporary directories for running HMRF.
HMRF_out &lt;- run_HMRF(sim_dat, xy, ztrue = info$z, R, case = 3, 
                     rep, usePCs = F, dosearchSEgenes = T)</code></pre>
<pre><code>
 external python path provided and will be used 
Consider to install these (optional) packages to run all possible Giotto commands for spatial analyses:  trendsceek multinet RTriangle FactoMiner
 Giotto does not automatically install all these packages as they are not absolutely required and this reduces the number of dependencies
 This is the single parameter version of binSpect
 1. matrix binarization complete 

 2. spatial enrichment test completed 

 3. (optional) average expression of high expressing cells calculated 

 4. (optional) number of high expressing cells calculated 
[1] &quot;/net/mulan/home/zlisph//softwares/miniconda3/bin/python /net/mulan/home/zlisph/R/x86_64-pc-linux-gnu-library/4.0/Giotto/python/reader2.py -l \&quot;/net/mulan/home/zlisph/BASS_pjt/2_simu/HMRF_tmp/case3/rep1/result_k4/spatial_cell_locations.txt\&quot; -g \&quot;/net/mulan/home/zlisph/BASS_pjt/2_simu/HMRF_tmp/case3/rep1/result_k4/spatial_genes.txt\&quot; -n \&quot;/net/mulan/home/zlisph/BASS_pjt/2_simu/HMRF_tmp/case3/rep1/result_k4/spatial_network.txt\&quot; -e \&quot;/net/mulan/home/zlisph/BASS_pjt/2_simu/HMRF_tmp/case3/rep1/result_k4/expression_matrix.txt\&quot; -o \&quot;/net/mulan/home/zlisph/BASS_pjt/2_simu/HMRF_tmp/case3/rep1/result_k4/result.spatial.zscore\&quot; -a test -k 4 -b 0 2 26 -t 1e-10 -z none -s 100 -i 100&quot;</code></pre>
<pre class="r"><code># ARI for spatial domain detection across different
# spatial interaction parameter betas.
HMRF_out</code></pre>
<pre><code>      0.0       2.0       4.0       6.0       8.0      10.0      12.0      14.0 
0.1508578 0.1508578 0.1576335 0.1648563 0.2533123 0.3469621 0.4017072 0.4408313 
     16.0      18.0      20.0      22.0      24.0      26.0      28.0      30.0 
0.4852844 0.4525879 0.4848594 0.5345169 0.5860648 0.6032793 0.6169157 0.6420285 
     32.0      34.0      36.0      38.0      40.0      42.0      44.0      46.0 
0.6339821 0.5739655 0.5729992 0.5880543 0.6063979 0.6116954 0.6033891 0.5984740 
     48.0      50.0 
0.5886572 0.5886572 </code></pre>
</div>
<div id="run-bayesspace" class="section level3">
<h3>Run BayesSpace</h3>
<p>Note that BayesSpace cannot identify any neighbors because it was developed for analyzing ordered lattice structure of spots from ST and 10x Visium platforms.</p>
<pre class="r"><code># Run BayesSpace
BayesSpace_out &lt;- run_BayesSpace(sim_dat, xy, ztrue = info$z, R)</code></pre>
<pre><code>Neighbors were identified for 0 out of 1049 spots.</code></pre>
<pre><code>Fitting model...</code></pre>
<pre><code>Calculating labels using iterations 1000 through 10000.</code></pre>
<pre class="r"><code># ARI for spatial domain detection
BayesSpace_out</code></pre>
<pre><code>[1] 0.1678712</code></pre>
</div>
<div id="run-spagcn" class="section level3">
<h3>Run SpaGCN</h3>
<p>Because SpaGCN was originally developed with Python, we wrapped up the code into a function and have it imported into R such that we can conveniently analyze the simulated data in R. The python function can be found at <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/code/">simulation</a>.</p>
<pre class="r"><code># Run SpaGCN
source_python(&quot;code/run_SpaGCN.py&quot;)
SpaGCN_out &lt;- run_SpaGCN(sim_dat, xy, info$z, R)
# ARI for spatial domain detection
SpaGCN_out</code></pre>
<pre><code>[1] 0.2550737</code></pre>
</div>
<div id="run-seurat" class="section level3">
<h3>Run Seurat</h3>
<p>Because Seurat uses a resolution parameter to indirectly determine the number of clusters, we run Seurat on a range of resolution parameters and identify the first value that resulted in the desired number of cell types.</p>
<pre class="r"><code># Run Seurat
resolutions &lt;- seq(0.3, 1.5, by = 0.1)
seu_out &lt;- seu_cluster(sim_dat, resolutions)
# ARI for cell type clustering and the number of identified clusters 
seu_out</code></pre>
<pre><code> [1] &quot;0.939455349549594,4&quot; &quot;0.948825866306038,4&quot; &quot;0.948825866306038,4&quot;
 [4] &quot;0.948825866306038,4&quot; &quot;0.951196703238985,4&quot; &quot;0.951196703238985,4&quot;
 [7] &quot;0.956035282669879,4&quot; &quot;0.956035282669879,4&quot; &quot;0.956035282669879,4&quot;
[10] &quot;0.958501798605292,4&quot; &quot;0.958501798605292,4&quot; &quot;0.84056426004742,5&quot; 
[13] &quot;0.762544765890606,6&quot;</code></pre>
</div>
<div id="run-sc3" class="section level3">
<h3>Run SC3</h3>
<p>Note that when the number of cells exceeds 5,000, SC3 randomly selects 5,000 cells for clustering, trains a support vector machine with the cluster labels obtained for the 5,000 cells, and then predict the cluster labels of the remaining cells.</p>
<pre class="r"><code># Run SC3
# Note SC3 algorithm is slow
sc3_out &lt;- sc3_cluster(sim_dat, C)</code></pre>
<pre><code>Setting SC3 parameters...</code></pre>
<pre><code>Calculating distances between the cells...</code></pre>
<pre><code>Performing transformations and calculating eigenvectors...</code></pre>
<pre><code>Performing k-means clustering...</code></pre>
<pre><code>Calculating consensus matrix...</code></pre>
<pre class="r"><code># ARI for cell type clustering
sc3_out</code></pre>
<pre><code>[1] 0.9670091</code></pre>
</div>
<div id="run-fict" class="section level3">
<h3>Run FICT</h3>
<p>Please install the software according to the <a href="https://github.com/haotianteng/FICT">FICT github</a> and modify the paths to the FICT software and a temporary folder accordingly in the <code>fict_cluster</code> function.</p>
<pre class="r"><code># Note that case and rep parameters below are used to generate a unique
# temporary directories for running FICT.
fict_out &lt;- fict_cluster(sim_dat, xy, C, case = 3, rep)
# ARI for cell type clustering
fict_out</code></pre>
<pre><code>[1] 0.9829696</code></pre>
</div>
</div>
<div id="multi-sample-simulation" class="section level2">
<h2>Multi-sample simulation</h2>
</div>
</div>
<div id="find-all-the-simulation-results" class="section level1">
<h1>Find all the simulation results</h1>
<p>Put the link to all the simulation results here later and plot the two main figures on simulation analysis.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] cowplot_1.1.1               reticulate_1.22            
 [3] splatter_1.16.1             gtools_3.9.2               
 [5] scater_1.16.2               forcats_0.5.0              
 [7] stringr_1.4.0               dplyr_1.0.8                
 [9] purrr_0.3.4                 readr_1.3.1                
[11] tidyr_1.1.1                 tibble_3.1.6               
[13] ggplot2_3.3.5               tidyverse_1.3.0            
[15] SC3_1.20.0                  Seurat_3.2.3               
[17] BayesSpace_1.2.1            SingleCellExperiment_1.14.1
[19] SummarizedExperiment_1.22.0 Biobase_2.48.0             
[21] GenomicRanges_1.44.0        GenomeInfoDb_1.24.2        
[23] IRanges_2.26.0              S4Vectors_0.30.2           
[25] BiocGenerics_0.38.0         MatrixGenerics_1.4.3       
[27] matrixStats_0.61.0          Giotto_1.0.4               
[29] BASS_1.0                    mclust_5.4.9               
[31] GIGrvg_0.5                  workflowr_1.7.0            

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3            scattermore_0.7          
  [3] maxLik_1.5-2              coda_0.19-4              
  [5] bit64_4.0.2               knitr_1.37               
  [7] irlba_2.3.3               DelayedArray_0.18.0      
  [9] data.table_1.14.2         rpart_4.1-15             
 [11] RCurl_1.98-1.5            doParallel_1.0.15        
 [13] generics_0.1.2            callr_3.7.0              
 [15] RSQLite_2.2.0             combinat_0.0-8           
 [17] RANN_2.6.1                proxy_0.4-26             
 [19] future_1.22.1             bit_4.0.4                
 [21] xml2_1.3.3                lubridate_1.7.9          
 [23] spatstat.data_2.1-0       httpuv_1.5.4             
 [25] assertthat_0.2.1          viridis_0.5.1            
 [27] xfun_0.29                 hms_0.5.3                
 [29] jquerylib_0.1.4           evaluate_0.15            
 [31] promises_1.1.1            DEoptimR_1.0-9           
 [33] fansi_1.0.2               readxl_1.3.1             
 [35] dbplyr_1.4.4              igraph_1.2.7             
 [37] DBI_1.1.1                 htmlwidgets_1.5.1        
 [39] ellipsis_0.3.2            backports_1.2.1          
 [41] deldir_1.0-6              sparseMatrixStats_1.4.2  
 [43] vctrs_0.3.8               ROCR_1.0-11              
 [45] abind_1.4-5               withr_2.4.3              
 [47] cachem_1.0.6              robustbase_0.93-9        
 [49] checkmate_2.0.0           sctransform_0.3.2        
 [51] scran_1.20.1              goftest_1.2-3            
 [53] cluster_2.1.2             lazyeval_0.2.2           
 [55] crayon_1.5.0              labeling_0.4.2           
 [57] edgeR_3.34.1              pkgconfig_2.0.3          
 [59] nlme_3.1-153              vipor_0.4.5              
 [61] rlang_1.0.1               globals_0.14.0           
 [63] lifecycle_1.0.1           miniUI_0.1.1.1           
 [65] sandwich_3.0-1            BiocFileCache_1.12.0     
 [67] modelr_0.1.8              rsvd_1.0.3               
 [69] cellranger_1.1.0          rprojroot_2.0.2          
 [71] polyclip_1.10-0           lmtest_0.9-38            
 [73] rngtools_1.5.2            Matrix_1.3-4             
 [75] Rhdf5lib_1.10.1           zoo_1.8-9                
 [77] reprex_0.3.0              beeswarm_0.4.0           
 [79] whisker_0.4               ggridges_0.5.3           
 [81] processx_3.5.2            pheatmap_1.0.12          
 [83] png_0.1-7                 viridisLite_0.4.0        
 [85] bitops_1.0-7              getPass_0.2-2            
 [87] KernSmooth_2.23-20        blob_1.2.1               
 [89] DelayedMatrixStats_1.14.3 doRNG_1.8.2              
 [91] parallelly_1.28.1         lpSolve_5.6.15           
 [93] beachmat_2.8.1            scales_1.1.1             
 [95] memoise_2.0.1             magrittr_2.0.2           
 [97] plyr_1.8.6                ica_1.0-2                
 [99] zlibbioc_1.34.0           DirichletReg_0.7-1       
[101] compiler_4.1.2            miscTools_0.6-26         
[103] dqrng_0.3.0               RColorBrewer_1.1-2       
[105] rrcov_1.6-0               fitdistrplus_1.1-6       
[107] cli_3.2.0                 XVector_0.32.0           
[109] listenv_0.8.0             patchwork_1.1.1          
[111] pbapply_1.5-0             ps_1.6.0                 
[113] Formula_1.2-4             MASS_7.3-54              
[115] mgcv_1.8-38               tidyselect_1.1.2         
[117] stringi_1.7.6             highr_0.9                
[119] yaml_2.3.5                BiocSingular_1.4.0       
[121] locfit_1.5-9.4            ggrepel_0.9.1            
[123] grid_4.1.2                sass_0.4.0               
[125] tools_4.1.2               future.apply_1.8.1       
[127] label.switching_1.8       rstudioapi_0.13          
[129] bluster_1.2.1             foreach_1.5.0            
[131] git2r_0.28.0              metapod_1.0.0            
[133] gridExtra_2.3             farver_2.1.0             
[135] Rtsne_0.15                digest_0.6.29            
[137] pracma_2.2.9              shiny_1.5.0              
[139] Rcpp_1.0.8                broom_0.7.10             
[141] scuttle_1.2.1             later_1.1.0.1            
[143] RcppAnnoy_0.0.19          WriteXLS_6.3.0           
[145] httr_1.4.2                Rdpack_2.1.2             
[147] colorspace_2.0-3          rvest_0.3.6              
[149] fs_1.5.2                  tensor_1.5               
[151] splines_4.1.2             uwot_0.1.10              
[153] statmod_1.4.36            spatstat.utils_2.2-0     
[155] xgboost_1.4.1.1           plotly_4.9.2.1           
[157] xtable_1.8-4              jsonlite_1.8.0           
[159] spatstat_1.64-1           R6_2.5.1                 
[161] pillar_1.7.0              htmltools_0.5.2          
[163] mime_0.12                 glue_1.6.2               
[165] fastmap_1.1.0             BiocParallel_1.22.0      
[167] BiocNeighbors_1.6.0       class_7.3-19             
[169] codetools_0.2-18          pcaPP_1.9-74             
[171] mvtnorm_1.1-3             utf8_1.2.2               
[173] lattice_0.20-45           bslib_0.3.1              
[175] curl_4.3.2                ggbeeswarm_0.6.0         
[177] leiden_0.3.9              survival_3.2-13          
[179] limma_3.48.3              rmarkdown_2.11           
[181] munsell_0.5.0             e1071_1.7-9              
[183] rhdf5_2.32.2              GenomeInfoDbData_1.2.6   
[185] iterators_1.0.13          smfishHmrf_0.1           
[187] haven_2.3.1               reshape2_1.4.4           
[189] gtable_0.3.0              rbibutils_2.2.4          </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
