<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zheng Li" />

<meta name="date" content="2022-02-24" />

<title>STARmap (Wang et al., 2018) analysis</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BASS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Introduction</a>
</li>
<li>
  <a href="about.html">Software tutorial</a>
</li>
<li>
  <a href="simu.html">Simulation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Real data analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="STARmap.html">STARmap (Wang et al., 2018)</a>
    </li>
    <li>
      <a href="MERFISH.html">MERFISH (Moffitt et al., 2018)</a>
    </li>
    <li>
      <a href="DLPFC.html">DLPFC (Maynard et al., 2021)</a>
    </li>
  </ul>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zhengli09//BASS-analysis">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">STARmap (Wang et al., 2018) analysis</h1>
<h4 class="author">Zheng Li</h4>
<h4 class="date">2022-02-24</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-02-27
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>BASS-analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed0code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(0)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed0code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(0)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzhengli09BASSanalysistree4ec8df1262087c7c78f682e56014a7087a756d4ftargetblank4ec8df1a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/zhengli09//BASS-analysis/tree/4ec8df1262087c7c78f682e56014a7087a756d4f" target="_blank">4ec8df1</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomzhengli09BASSanalysistree4ec8df1262087c7c78f682e56014a7087a756d4ftargetblank4ec8df1a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/zhengli09//BASS-analysis/tree/4ec8df1262087c7c78f682e56014a7087a756d4f" target="_blank">4ec8df1</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/STARmap.Rmd</code>) and HTML (<code>docs/STARmap.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zhengli09//BASS-analysis/blob/4ec8df1262087c7c78f682e56014a7087a756d4f/analysis/STARmap.Rmd" target="_blank">4ec8df1</a>
</td>
<td>
zhengli09
</td>
<td>
2022-02-27
</td>
<td>
Publish the introduction, simulation analysis and STARmap analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Here, we apply BASS to analyze the STARmap mPFC (mouse medial prefrontal cortex) data from <a href="https://www.science.org/doi/10.1126/science.aat5691">Wang et al. 2018</a>. For single-sample analysis, we focus on the tissue section BZ5. For multi-sample analysis, we focus on tissue sections BZ5, BZ9 and BZ14 that were measured on the mPFC region of different mice. The original data can be downloaded from <a href="https://www.starmapresources.org/data">here</a>. We excluded cells that are annotated to be “NA” class by the original study as they were not confidently identified to be any cell type. Finally, we obtained the same set of 166 genes measured on 1,049 (BZ5), 1,053 (BZ9), and 1,088 (BZ14) cells along with their centroid coordinates for the following analysis. The processed data can be download from the <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/data">data directory</a>. For detailed usage of all the functions, please refer to the software tutorial section.</p>
<pre class="r"><code>library(BASS)
load(&quot;data/starmap_mpfc.RData&quot;)</code></pre>
</div>
<div id="single-sample-analysis" class="section level1">
<h1>Single-sample analysis</h1>
<div id="load-data-and-set-hyper-parameters" class="section level2">
<h2>Load data and set hyper-parameters</h2>
<pre class="r"><code># We focus on the tissue section BZ5 in the single-sample analysis
cnts &lt;- starmap_cnts[&quot;20180417_BZ5_control&quot;] # 166x1049 count data
info &lt;- starmap_info[&quot;20180417_BZ5_control&quot;]
xy &lt;- lapply(info, function(info.i){
  as.matrix(info.i[, c(&quot;x&quot;, &quot;y&quot;)]) # 1049x2 spatial coordiantes
})
cnts[[1]][1:5,1:5]</code></pre>
<pre><code>        69x4486 93x1063 143x3445 88x4092 120x4293
Acss1         0       0        0       0        0
Adcyap1       0       3        0       0        0
Adgrl2        0       0       16       1        7
Aqp4          0       0        8      11       32
Arc           3       3        1       5        0</code></pre>
<pre class="r"><code>head(xy[[1]])</code></pre>
<pre><code>                 x        y
69x4486   68.86374 4486.029
93x1063   92.84376 1062.676
143x3445 142.82557 3444.635
88x4092   88.18908 4092.441
120x4293 119.93251 4293.439
104x2648 103.63420 2647.580</code></pre>
<pre class="r"><code># hyper-parameters
C &lt;- 15 # number of cell types
R &lt;- 4 # number of spatial domains</code></pre>
</div>
<div id="run-bass" class="section level2">
<h2>Run BASS</h2>
<pre class="r"><code>set.seed(0)
# Set up BASS object
BASS &lt;- createBASSObject(cnts, xy, C = C, R = R, 
  beta_est_approach = &quot;ACCUR_EST&quot;, 
  burn_in = 10000, samples = 10000)</code></pre>
<pre><code>***************************************
  Bayesian Analytics for Spatial Segmentation (BASS)
  Authors: Zheng Li, Xiang Zhou
  Affiliate: Department of Biostatistics, University of Michigan
  INPUT INFO:
    - Number of samples: 1 
    - Number of spots/cells: 1049 
    - Number of genes: 166 
    - Potts interaction parameter estimation approach: ACCUR_EST 
    - Variance-covariance structure of gene expression features: EEE 
  To list all hyper-parameters, Type listAllHyper(BASS_object)
***************************************</code></pre>
<pre class="r"><code># Data pre-processing:
# 1.Library size normalization followed with a log2 transformation
# 2.Dimension reduction with PCA after standardizing all the genes
BASS &lt;- BASS.preprocess(BASS, doLogNormalize = TRUE, 
  doPCA = TRUE, scaleFeature = TRUE, nPC = 20)</code></pre>
<pre><code>***** Log-normalize gene expression data *****
***** Exclude genes with 0 expression *****
***** Reduce data dimension with PCA *****</code></pre>
<pre class="r"><code># Run BASS algorithm
BASS &lt;- BASS.run(BASS)</code></pre>
<pre class="r"><code># post-process posterior samples:
# 1.Adjust for label switching with the ECR-1 algorithm
# 2.Summarize the posterior samples to obtain the cell type labels, spatial 
#   domain labels, and cell type proportion matrix estiamte
BASS &lt;- BASS.postprocess(BASS)</code></pre>
<pre><code>
    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                58.144               Converged (3 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 61.9 seconds. 

    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                21.093               Converged (2 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 22.8 seconds. </code></pre>
<pre class="r"><code>clabels &lt;- BASS@res_postprocess$c_ls # cell type clusters
zlabels &lt;- BASS@res_postprocess$z_ls # spatial domain labels
pi_est &lt;- BASS@res_postprocess$pi_ls # cell type composition matrix</code></pre>
</div>
<div id="annotate-cell-types" class="section level2">
<h2>Annotate cell types</h2>
<pre class="r"><code>library(Seurat)
library(dplyr)</code></pre>
<pre class="r"><code>l &lt;- 1 # tissue section number
# Perform DE analysis with Seurat
seu_obj &lt;- CreateSeuratObject(counts = cnts[[l]], min.cells = 1)
seu_obj &lt;- NormalizeData(seu_obj)
seu_obj &lt;- ScaleData(seu_obj, features = rownames(seu_obj))</code></pre>
<pre><code>Centering and scaling data matrix</code></pre>
<pre class="r"><code>seu_obj &lt;- RunPCA(seu_obj, features = rownames(seu_obj), verbose = F)
Idents(seu_obj) &lt;- factor(clabels[[l]])
markers &lt;- FindAllMarkers(seu_obj, only.pos = T, 
  min.pct = 0, logfc.threshold = 0, verbose = F)
top5 &lt;- markers %&gt;%
  group_by(cluster) %&gt;%
  top_n(n = 5, wt = avg_logFC)
subset(top5, cluster == 7) # Markers of eL2/3 cells</code></pre>
<pre><code># A tibble: 5 × 7
# Groups:   cluster [1]
     p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene 
     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;
1 3.13e-62      2.41 0.892 0.181  5.20e-60 7       Cux2 
2 3.42e-29      1.41 0.753 0.254  5.68e-27 7       Nos1 
3 5.14e-29      1.63 0.634 0.167  8.53e-27 7       Synpr
4 2.70e-21      1.15 0.914 0.546  4.48e-19 7       Nov  
5 1.57e-19      1.15 0.892 0.611  2.61e-17 7       Otof </code></pre>
<pre class="r"><code># By checking the top DE genes of each cell type cluster, we annotate 
# specific cell types for each cluster
cTypes &lt;- c(&quot;eL5d&quot;, &quot;eL6&quot;, &quot;OPC&quot;, &quot;eL5c&quot;, &quot;eL5a&quot;, 
  &quot;Astro&quot;, &quot;eL2/3&quot;, &quot;Reln&quot;, &quot;SST-Lhx6&quot;, &quot;eL5b&quot;, 
  &quot;Smc&quot;, &quot;Micro&quot;, &quot;Smc-Astro&quot;, &quot;Oligo&quot;, &quot;VIP&quot;)
clabels_l &lt;- factor(clabels[[l]])
levels(clabels_l) &lt;- cTypes</code></pre>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>You can refer to <a href="https://github.com/zhengli09/BASS-Analysis/tree/main/code/viz.R">visualization</a> for some useful plotting functions or you can write your own code for plotting.</p>
<pre class="r"><code>library(ggplot2)
source(&quot;code/viz.R&quot;)
# Spatial domains
zlabels_l &lt;- factor(zlabels[[l]])
levels(zlabels_l) &lt;- c(&quot;L2/3&quot;, &quot;L1&quot;, &quot;L6&quot;, &quot;L5&quot;)
zlabels_l &lt;- factor(zlabels_l, c(&quot;L1&quot;, &quot;L2/3&quot;, &quot;L5&quot;, &quot;L6&quot;))
plotClusters(xy[[l]], labels = zlabels_l, title = &quot;Spatial domains&quot;) +
  theme(legend.position = &quot;bottom&quot;)</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Cell types
library(gghighlight)
clabels_l &lt;- factor(clabels_l, levels = c(
  &quot;eL2/3&quot;, &quot;eL5a&quot;, &quot;eL5b&quot;, &quot;eL5c&quot;, &quot;eL5d&quot;, &quot;eL6&quot;,
  &quot;Reln&quot;, &quot;VIP&quot;, &quot;SST-Lhx6&quot;,
  &quot;Oligo&quot;, &quot;Smc&quot;, &quot;Astro&quot;, &quot;Smc-Astro&quot;, &quot;Micro&quot;, &quot;OPC&quot;)
  )
plotCellTypes(xy, clabels_l, ncol = 5) + 
  ggtitle(&quot;Spatial distribution of cell types&quot;) +
  theme(plot.title = element_text(hjust = 0.5))</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-8-1.png" width="816" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Cell type compositions in each spatial domain
pi_est &lt;- BASS@res_postprocess$pi_ls
pi_est &lt;- pi_est[match(levels(clabels_l), cTypes), c(2, 1, 4, 3)]
rownames(pi_est) &lt;- levels(clabels_l)
colnames(pi_est) &lt;- levels(zlabels_l)
plotCellTypeComp(pi_est)</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-9-1.png" width="288" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="multi-sample-analysis" class="section level1">
<h1>Multi-sample analysis</h1>
<div id="load-data-and-set-hyper-parameters-1" class="section level2">
<h2>Load data and set hyper-parameters</h2>
<pre class="r"><code># We use all the three tissue sections: BZ5, BZ9, and BZ14 in 
# the multi-sample analysis
xy &lt;- lapply(starmap_info, function(info.i){
  as.matrix(info.i[, c(&quot;x&quot;, &quot;y&quot;)]) # spatial coordiantes
})
cnts &lt;- starmap_cnts # count matrices for three tissue sections
# hyper-parameters
C &lt;- 15 # number of cell types
R &lt;- 4 # number of spatial domains</code></pre>
</div>
<div id="run-bass-1" class="section level2">
<h2>Run BASS</h2>
<pre class="r"><code>set.seed(0)
# Set up BASS object
BASS &lt;- createBASSObject(cnts, xy, C = C, R = R, 
  beta_est_approach = &quot;ACCUR_EST&quot;, 
  burn_in = 10000, samples = 10000)</code></pre>
<pre><code>***************************************
  Bayesian Analytics for Spatial Segmentation (BASS)
  Authors: Zheng Li, Xiang Zhou
  Affiliate: Department of Biostatistics, University of Michigan
  INPUT INFO:
    - Number of samples: 3 
    - Number of spots/cells: 1049 1053 1088 
    - Number of genes: 166 
    - Potts interaction parameter estimation approach: ACCUR_EST 
    - Variance-covariance structure of gene expression features: EEE 
  To list all hyper-parameters, Type listAllHyper(BASS_object)
***************************************</code></pre>
<pre class="r"><code># Data pre-processing:
# In addition to the log-normalization and dimension reduction with
# PCA, we also conduct a batch effect adjustment using the Harmony
# package.
BASS &lt;- BASS.preprocess(BASS, doLogNormalize = TRUE, 
  doPCA = TRUE, scaleFeature = TRUE, nPC = 20)</code></pre>
<pre><code>***** Log-normalize gene expression data *****
***** Exclude genes with 0 expression *****
***** Reduce data dimension with PCA *****
***** Correct batch effect with Harmony *****</code></pre>
<pre class="r"><code># Run BASS algorithm
BASS &lt;- BASS.run(BASS)</code></pre>
<pre class="r"><code># post-process posterior samples:
BASS &lt;- BASS.postprocess(BASS)</code></pre>
<pre><code>
    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                118.055              Converged (3 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 125.5 seconds. 

    ......................................................................................
    . Method                         Time (sec)           Status                         . 
    ......................................................................................
    . ECR-ITERATIVE-1                53.527               Converged (2 iterations)       . 
    ......................................................................................

    Relabelling all methods according to method ECR-ITERATIVE-1 ... done!
    Retrieve the 1 permutation arrays by typing:
        [...]$permutations$&quot;ECR-ITERATIVE-1&quot;
    Retrieve the 1 best clusterings: [...]$clusters
    Retrieve the 1 CPU times: [...]$timings
    Retrieve the 1 X 1 similarity matrix: [...]$similarity
    Label switching finished. Total time: 59.3 seconds. </code></pre>
<pre class="r"><code>clabels &lt;- BASS@res_postprocess$c_ls # cell type clusters
zlabels &lt;- BASS@res_postprocess$z_ls # spatial domain labels
pi_est &lt;- BASS@res_postprocess$pi_ls # cell type composition matrix</code></pre>
</div>
<div id="annotate-cell-types-1" class="section level2">
<h2>Annotate cell types</h2>
<pre class="r"><code># Perform DE analysis with Seurat
cnts_all &lt;- do.call(cbind, cnts)
seu_obj &lt;- CreateSeuratObject(counts = cnts_all, min.cells = 1)
seu_obj &lt;- NormalizeData(seu_obj)
seu_obj &lt;- ScaleData(seu_obj, features = rownames(seu_obj))</code></pre>
<pre><code>Centering and scaling data matrix</code></pre>
<pre class="r"><code>seu_obj &lt;- RunPCA(seu_obj, features = rownames(seu_obj), verbose = F)
Idents(seu_obj) &lt;- factor(unlist(clabels))
markers &lt;- FindAllMarkers(seu_obj, only.pos = T, 
  min.pct = 0, logfc.threshold = 0, verbose = F)
top5 &lt;- markers %&gt;%
  group_by(cluster) %&gt;%
  top_n(n = 5, wt = avg_logFC)
subset(top5, cluster == 7) # Markers of eL6a cells</code></pre>
<pre><code># A tibble: 5 × 7
# Groups:   cluster [1]
      p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene 
      &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;
1 4.90e-259      2.23 0.746 0.122 8.13e-257 7       Syt6 
2 6.11e-252      1.65 0.983 0.485 1.01e-249 7       Pcp4 
3 2.63e-207      1.70 0.878 0.287 4.37e-205 7       Sla  
4 8.65e-127      1.23 0.589 0.146 1.44e-124 7       Foxp2
5 6.80e- 44      1.35 0.171 0.028 1.13e- 41 7       Rprm </code></pre>
<pre class="r"><code>subset(top5, cluster == 3) # Markers of eL6b cells</code></pre>
<pre><code># A tibble: 5 × 7
# Groups:   cluster [1]
     p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene 
     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;
1 4.31e-86      2.74 0.947 0.324  7.15e-84 3       Ctgf 
2 2.85e-46      1.23 0.977 0.57   4.73e-44 3       Pcp4 
3 3.14e-27      1.43 0.737 0.357  5.21e-25 3       Cplx3
4 4.64e-27      1.21 0.632 0.238  7.70e-25 3       Nr4a2
5 8.89e-23      1.93 0.188 0.028  1.48e-20 3       Obox3</code></pre>
<pre class="r"><code># By checking the top DE genes of each cell type cluster, we annotate 
# specific cell types for each cluster
cTypes &lt;- c(
  &quot;Lhx6&quot;, &quot;Micro&quot;, &quot;eL6b&quot;, &quot;eL5d&quot;, &quot;eL2/3&quot;, 
  &quot;Oligo&quot;, &quot;eL6a&quot;, &quot;eL5c&quot;, &quot;Smc&quot;, &quot;eL5a&quot;, 
  &quot;eL5b&quot;, &quot;Astro&quot;, &quot;SST&quot;, &quot;VIP&quot;, &quot;Reln&quot;)
clabels &lt;- lapply(clabels, function(clabels.l){
  clabels.l &lt;- factor(clabels.l)
  levels(clabels.l) &lt;- cTypes
  clabels.l &lt;- factor(clabels.l, levels = c(
    &quot;eL2/3&quot;, &quot;eL5a&quot;, &quot;eL5b&quot;, &quot;eL5c&quot;, &quot;eL5d&quot;, 
    &quot;eL6a&quot;, &quot;eL6b&quot;, &quot;Reln&quot;, &quot;VIP&quot;, &quot;SST&quot;,&quot;Lhx6&quot;,
    &quot;Oligo&quot;, &quot;Smc&quot;, &quot;Astro&quot;, &quot;Micro&quot;
  ))
})</code></pre>
</div>
<div id="visualization-1" class="section level2">
<h2>Visualization</h2>
<pre class="r"><code># Spatial domains
library(cowplot)
zlabels &lt;- lapply(zlabels, function(zlabels.l){
  zlabels.l &lt;- factor(zlabels.l)
  levels(zlabels.l) &lt;- c(&quot;L6&quot;, &quot;L5&quot;, &quot;L2/3&quot;, &quot;L1&quot;)
  zlabels.l &lt;- factor(zlabels.l, levels = c(&quot;L1&quot;, &quot;L2/3&quot;, &quot;L5&quot;, &quot;L6&quot;))
})
plist &lt;- lapply(1:length(zlabels), function(i){
  title.i &lt;- c(&quot;BZ5&quot;, &quot;BZ9&quot;, &quot;BZ14&quot;)[i]
  plotClusters(xy[[i]], labels = zlabels[[i]], title = title.i)
})
plot_grid(plotlist = plist, ncol = 3)</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-14-1.png" width="864" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Cell types
library(RColorBrewer)
getPalette &lt;- colorRampPalette(brewer.pal(8, &quot;Dark2&quot;))
plist &lt;- lapply(1:length(clabels), function(i){
  title.i &lt;- c(&quot;BZ5&quot;, &quot;BZ9&quot;, &quot;BZ14&quot;)[i]
  plotClusters(xy[[i]], labels = clabels[[i]], title.i) +
  scale_colour_manual(&quot;Cell types&quot;, values = getPalette(15))
})
legend &lt;- get_legend(plist[[1]] + theme(legend.position = &quot;bottom&quot;))
plist &lt;- append(plist, list(legend) , 3)
plot_grid(plotlist = plist, ncol = 2)</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-15-1.png" width="816" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Spatial distribution of cell types on tissue section BZ5
l &lt;- 1
plotCellTypes(xy[[l]], clabels[[l]], ncol = 5) +
  scale_colour_manual(&quot;Cell types&quot;, values = getPalette(15))</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-16-1.png" width="816" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Cell type compositions in each spatial domain
pi_est &lt;- BASS@res_postprocess$pi_ls
pi_est &lt;- pi_est[match(levels(clabels[[l]]), cTypes), c(4, 3, 2, 1)]
rownames(pi_est) &lt;- levels(clabels[[l]])
colnames(pi_est) &lt;- levels(zlabels[[l]])
plotCellTypeComp(pi_est)</code></pre>
<p><img src="figure/STARmap.Rmd/unnamed-chunk-17-1.png" width="288" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] RColorBrewer_1.1-2 cowplot_1.1.1      gghighlight_0.3.2  ggplot2_3.3.5     
 [5] dplyr_1.0.8        Seurat_3.2.3       BASS_1.0           mclust_5.4.9      
 [9] GIGrvg_0.5         workflowr_1.7.0   

loaded via a namespace (and not attached):
  [1] plyr_1.8.6                  igraph_1.2.7               
  [3] lazyeval_0.2.2              splines_4.1.2              
  [5] BiocParallel_1.22.0         listenv_0.8.0              
  [7] scattermore_0.7             GenomeInfoDb_1.24.2        
  [9] scater_1.16.2               digest_0.6.29              
 [11] htmltools_0.5.2             viridis_0.5.1              
 [13] fansi_1.0.2                 magrittr_2.0.2             
 [15] tensor_1.5                  cluster_2.1.2              
 [17] ROCR_1.0-11                 limma_3.48.3               
 [19] globals_0.14.0              matrixStats_0.61.0         
 [21] lpSolve_5.6.15              colorspace_2.0-3           
 [23] blob_1.2.1                  ggrepel_0.9.1              
 [25] xfun_0.29                   callr_3.7.0                
 [27] crayon_1.5.0                RCurl_1.98-1.5             
 [29] jsonlite_1.8.0              spatstat.data_2.1-0        
 [31] spatstat_1.64-1             survival_3.2-13            
 [33] zoo_1.8-9                   glue_1.6.2                 
 [35] polyclip_1.10-0             gtable_0.3.0               
 [37] zlibbioc_1.34.0             XVector_0.32.0             
 [39] leiden_0.3.9                DelayedArray_0.18.0        
 [41] BiocSingular_1.4.0          future.apply_1.8.1         
 [43] SingleCellExperiment_1.14.1 BiocGenerics_0.38.0        
 [45] abind_1.4-5                 scales_1.1.1               
 [47] DBI_1.1.1                   miniUI_0.1.1.1             
 [49] Rcpp_1.0.8                  viridisLite_0.4.0          
 [51] xtable_1.8-4                reticulate_1.22            
 [53] rsvd_1.0.3                  stats4_4.1.2               
 [55] htmlwidgets_1.5.1           httr_1.4.2                 
 [57] ellipsis_0.3.2              ica_1.0-2                  
 [59] farver_2.1.0                pkgconfig_2.0.3            
 [61] uwot_0.1.10                 deldir_1.0-6               
 [63] sass_0.4.0                  utf8_1.2.2                 
 [65] labeling_0.4.2              tidyselect_1.1.2           
 [67] rlang_1.0.1                 reshape2_1.4.4             
 [69] later_1.1.0.1               munsell_0.5.0              
 [71] tools_4.1.2                 cli_3.2.0                  
 [73] generics_0.1.2              ggridges_0.5.3             
 [75] evaluate_0.15               stringr_1.4.0              
 [77] fastmap_1.1.0               goftest_1.2-3              
 [79] yaml_2.3.5                  processx_3.5.2             
 [81] knitr_1.37                  fs_1.5.2                   
 [83] fitdistrplus_1.1-6          purrr_0.3.4                
 [85] RANN_2.6.1                  nlme_3.1-153               
 [87] pbapply_1.5-0               future_1.22.1              
 [89] sparseMatrixStats_1.4.2     whisker_0.4                
 [91] mime_0.12                   compiler_4.1.2             
 [93] rstudioapi_0.13             beeswarm_0.4.0             
 [95] plotly_4.9.2.1              png_0.1-7                  
 [97] spatstat.utils_2.2-0        tibble_3.1.6               
 [99] bslib_0.3.1                 stringi_1.7.6              
[101] highr_0.9                   ps_1.6.0                   
[103] lattice_0.20-45             Matrix_1.3-4               
[105] vctrs_0.3.8                 pillar_1.7.0               
[107] lifecycle_1.0.1             combinat_0.0-8             
[109] lmtest_0.9-38               jquerylib_0.1.4            
[111] RcppAnnoy_0.0.19            BiocNeighbors_1.6.0        
[113] data.table_1.14.2           bitops_1.0-7               
[115] irlba_2.3.3                 httpuv_1.5.4               
[117] patchwork_1.1.1             GenomicRanges_1.44.0       
[119] R6_2.5.1                    promises_1.1.1             
[121] KernSmooth_2.23-20          gridExtra_2.3              
[123] vipor_0.4.5                 IRanges_2.26.0             
[125] parallelly_1.28.1           codetools_0.2-18           
[127] MASS_7.3-54                 assertthat_0.2.1           
[129] SummarizedExperiment_1.22.0 rprojroot_2.0.2            
[131] withr_2.4.3                 label.switching_1.8        
[133] sctransform_0.3.2           harmony_0.1.0              
[135] S4Vectors_0.30.2            GenomeInfoDbData_1.2.6     
[137] mgcv_1.8-38                 parallel_4.1.2             
[139] rpart_4.1-15                grid_4.1.2                 
[141] tidyr_1.1.1                 rmarkdown_2.11             
[143] DelayedMatrixStats_1.14.3   MatrixGenerics_1.4.3       
[145] Rtsne_0.15                  git2r_0.28.0               
[147] getPass_0.2-2               Biobase_2.48.0             
[149] shiny_1.5.0                 ggbeeswarm_0.6.0           </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
